### ISSUE TITLE
Support multi-theme builds

### ISSUE THREAD

**Issue by nklayman** [quasarframework/vue-cli-plugin-quasar#7]
Instead of requiring the user to change the theme manually in `vue.config.js`, there should be an option for building both themes with separate output folders. You should register a `build:quasar` command that triggers the build twice with the different themes. I will make a PR for this if requested.

**Comment by rstoenescu** [comment]
Hi,

Yes, please do a PR. Thank you.

**Comment by nklayman** [comment]
I added the `build:quasar` commands, [repo is here](https://github.com/nklayman/vue-cli-plugin-quasar). It outputs `dist/quasar-theme-mat` and `dist/quasar-theme-ios`. I haven't opened a PR as I think the below option is better, but I can if you prefer this command.

I was thinking that instead of just adding the `build:quasar` command for multi theme builds, Quasar could replace the default build command. Then, it could take arguments like `--mat` and `--ios` for themes and `--electron` and `--cordova` for targets (powered by [vue-cli-plugin-electron-builder](https://nklayman.github.io/vue-cli-plugin-electron-builder) and [vue-cli-plugin-cordova](https://github.com/dekimasoon/vue-cli-plugin-cordova)). It could then rival the power of quasar-cli while keeping the flexibility of vue-cli. What do you think? I would be open to working to implement this feature into vue-cli-plugin-quasar.

**Comment by rstoenescu** [comment]
Rivaling Quasar CLI is **very hard** due to the numerous things that Quasar CLI offers under the covers, but please do a try. If everything checks out, we'll then merge your PR. Thank you for your efforts!

**Comment by nklayman** [comment]
Should you create an `@qusar/vue-cli-plugin-cordova` instead of using the one I mentioned before? I know very little about Cordova so I'm not sure if it is a good plugin. As for electron, I made vue-cli-plugin-electron-builder, so I know that's good. Should Quasar also take over the `serve` command so that `--electron`, `--cordova`, and `--theme` would work there too?

**Comment by nklayman** [comment]
You'll probably want to create (or I could) a custom `@quasar/vue-cli-plugin-cordova` as the existing one is pretty basic and even installs vuetify by default,  LoL 😂. All jokes aside, I'm making progress on web/electron targets, and should have a working prototype within the week. I'm using TDDish, so it will have 100% coverage right out of the gate, and I will also set up some e2e tests. As for the SSR target, what about [this plugin](https://github.com/Akryum/vue-cli-plugin-ssr)? It seems pretty solid. Thanks for all your great work with Quasar, and I'm so glad to be helping out!

**Comment by nklayman** [comment]
@rstoenescu good news! I have support for both electron and web targets now! To use electron, you will have to use an unpublished version of vue-cli-plugin-electron-builder:
```
# Install from github repo
yarn add -D nklayman/vue-cli-plugin-electron-builder
# Invoke generator
vue invoke electron-builder
```

 I'll list some of the features below:

### Multi-theme and multi-platform builds

The plugin now overwrites the default build command. This means you can run commands like the following:
`yarn build --web --electron --mat --ios`
This will build for Electron and SPA, with both themes.

### Nested args (broken)

Currently this feature is broken until [yargs](https://github.com/yargs/yargs) figures out [this issue](https://github.com/yargs/yargs/issues/360). However, as soon as it is fixed I believe they will work.

You can nest arguments to further customize the build. For example, to build for electron with the ios theme and for web with the mat theme:
`yarn build --mat "--electron" --ios "--web"`

### Custom configurations

Since nested args can be annoying to type multiple times, you can use custom configurations. You can also use them to see how nested args will work until the yargs issue is solved. They are defined in `vue.config.js` like so:

```javascript
// vue.config.js

module.exports = {
  pluginOptions: {
    quasar: {
      configurations: {
        // You can nave multiple configs
        production: {
          // Works like nested args, except in JSON format
          mat: {
            electron: {
              // When building material theme for electron, build as an unpacked app for windows
              windows: 'dir'
            }
          },
          ios: {
            electron: {
              // Use ios theme for linux builds
              linux: 'dir'
            }
          },
          // Build for web with both themes
          web: true
        }
      }
    }
  },
  transpileDependencies: [/[\\\/]node_modules[\\\/]quasar-framework[\\\/]/]
}
```

Then use it via `yarn build --config production`.

### Testing

All the features I added are tested via Jest

### Try it yourself

Check out the fork [here](https://github.com/nklayman/vue-cli-plugin-quasar/tree/overwrite-default-commands). Make sure to switch to the `overwrite-default-commands` branch. There  may be bugs, as this is just an early prototype. I will add SSR support soon, but the cordova plugin needs to be rewritten as mentioned above. I still need to implement a serve command as well.

**Comment by nklayman** [comment]
I added a `serve:quasar` command. It works like so:

`yarn serve (-t|--theme) (mat|ios) (-p|--platform) (web|electron)`

**Comment by m0dch3n** [comment]
@nklayman 
Dekimasoon's Cordova Plugin was replaced by mine https://github.com/m0dch3n/vue-cli-plugin-cordova/

He thought my approach is better. In fact, my approach is based on his, and quasar-cli.

It's adding cordova the same way, then quasar-cli does. However it doesn't support $q things from quasar, but it already sets `process.env.CORDOVA_PLATFORM` to `android` or `ios`

What my plugin is missing, to support quasar completly, are the $q specific things

@rstoenescu 
Can you confirm, that only the following parts, are important ? 
```javascript
if (process.env.CORDOVA_PLATFORM) {
  Vue.prototype.$q.is.cordova = true 

  document.addEventListener('deviceready', () => {
    Vue.prototype.$q.cordova = window.cordova
  })
}
```

$q.is.cordova is important i.e. because of https://github.com/quasarframework/quasar/blob/472983e7d1419f17dc1986ada9076275ce393c11/src/history.js#L9

My idea, to integrate the cordova part to this plugin, would be, to prompt the user, if he wants cordova support, and if so, run `vue add cordova` and add the important $q things...

Caveat, a user can't easily add cordova later, so a second idea, would be, to create a new plugin, and when you decide to add cordova to quasar, you run `vue add quasar-cordova`

What do you think is better ?

**Comment by nklayman** [comment]
Not to shoot down your plugin (I've tried it and it's pretty good), but I'm thinking of not supporting cordova. Instead, [Capacitor](https://capacitor.ionicframework.com/) seems like a better idea. I was having issues with cordova, so I gave capacitor a shot, and I loved it. It is under active development by the Ionic team, and seems to be passing Cordova as the go-to option for making a mobile app out of a website. While Cordova has a complex file structure with two `package.json` files, capacitor only has a `capacitor.config.json`, and a single folder for `android` and `ios`. Instead of handling the native build step itself, it lets you run the build through Android Studio/xCode (which I had to do anyway for Cordova). It also has it's own set of functions for accessing native api's, that works on web, android, and ios. I wouldn't mind making a `vue-cli-plugin-capacitor` (I'm planning on making one anyway) that could integrate with `vue-cli-plugin-quasar`.

TL;DR: Cordova has some issues that are fixed by Capacitor. I think we should use it instead of Cordova.

As a side note, I recently submitted [an issue](https://github.com/ionic-team/capacitor/issues/849) suggesting a refactor of capacitor's layout.

**Comment by m0dch3n** [comment]
Yeah, no problem, I'm always open for new things to try ;)

Nevertheless quasar currently uses cordova, and to get it completly working (backbutton handling, deviceready handling etc) it will be harder to go with Capacitor IMO.

My plugin was mainly created to add cordova, to a vue cli app. That's why it doesn't has any quasar related things. I think, it's better, if I create a new issue therefore, and stay here on the main topic, of letting theming be a possibility...

**Comment by nklayman** [comment]
Nothing stops us from supporting both. You can go ahead and clone [my repo](https://github.com/nklayman/vue-cli-plugin-quasar) and add Cordova support. I'll happily merge the PR!

**Comment by m0dch3n** [comment]
I think it's more logic to directly do a PR here, once I have a response what @rstoenescu says about questions or idea...

**Comment by nklayman** [comment]
The reason I suggest doing the PR in my repo is I already have the multi-target build feature, it just doesn't use your plugin yet, instead saying 'Cordova is not yet supported'. I also have an initial version of [vue-cli-plugin-capacitor](https://github.com/nklayman/vue-cli-plugin-capacitor) (no readme yet).

**Comment by nklayman** [comment]
@rstoenescu bump
